src/main/java/uk/ac/ebi/atlas/search/GeneSearchDao.java

```diff
@@ -138,58 +143,53 @@ public GeneSearchDao(NamedParameterJdbcTemplate namedParameterJdbcTemplate,
         );
     }
 
-    // Returns inferred cell types and organism parts for each experiment accession
-    public Map<String, Map<String, List<String>>>
-    getFacets(List<String> cellIds, SingleCellAnalyticsSchemaField... singleCellAnalyticsSchemaFields) {
-        List<SingleCellAnalyticsSchemaField> subFacetFields = Arrays.asList(singleCellAnalyticsSchemaFields);
+    // Returns all the metadata values for each experiment accession, given a subset of metadata types
+    public Map<String, Map<String, List<String>>> getFacets(List<String> cellIds, String... metadataTypes) {
+        List<String> facetFieldValues = Arrays.asList(metadataTypes);
+
+        SolrJsonFacetBuilder<SingleCellAnalyticsCollectionProxy> characteristicValueFacet =
+                new SolrJsonFacetBuilder<SingleCellAnalyticsCollectionProxy>()
+                        .setFacetField(CHARACTERISTIC_VALUE)
+                        .isNestedFacet(true);
+
+        SolrJsonFacetBuilder<SingleCellAnalyticsCollectionProxy> characteristicNameFacet =
+                new SolrJsonFacetBuilder<SingleCellAnalyticsCollectionProxy>()
+                        .setFacetField(CHARACTERISTIC_NAME)
+                        .addSubFacets(ImmutableList.of(characteristicValueFacet))
+                        .isNestedFacet(true);
+
+        SolrJsonFacetBuilder<SingleCellAnalyticsCollectionProxy> facets =
+                new SolrJsonFacetBuilder<SingleCellAnalyticsCollectionProxy>()
+                        .setFacetField(EXPERIMENT_ACCESSION)
+                        .addSubFacets(ImmutableList.of(characteristicNameFacet));
 
         SolrQueryBuilder<SingleCellAnalyticsCollectionProxy> queryBuilder =
                 new SolrQueryBuilder<SingleCellAnalyticsCollectionProxy>()
+                        .addQueryFieldByTerm(CHARACTERISTIC_NAME, facetFieldValues)
                         .addQueryFieldByTerm(CELL_ID, cellIds)
-                        .setFacetField(EXPERIMENT_ACCESSION)
-                        .setSubFacetList(subFacetFields)
+                        .setFacets(facets)
                         .setRows(0);
 
         List<SimpleOrderedMap> resultsByExperiment =
-                extractSimpleOrderdMaps(
+                extractSimpleOrderedMaps(
                         singleCellAnalyticsCollectionProxy
                                 .query(queryBuilder)
                                 .getResponse()
                                 .findRecursive("facets", EXPERIMENT_ACCESSION.name(), "buckets"));
 
-
-        return resultsByExperiment == null ?
+        return isEmpty(resultsByExperiment) ?
                 emptyMap() :
```

Comment:
Again, if `resultsByExperiment` is empty, so will be the stream and the result of the return. You may want to add a test that checks that specific condition if there is none.

Reply:
I think the point was to exit early for efficiency's sake? This is a check [you added](https://github.com/ebi-gene-expression-group/atlas/blame/dev/sc/src/main/java/uk/ac/ebi/atlas/search/GeneSearchDao.java), but the `resultsByExperiment` couldn't be null I think, which is why I replaced it with `isEmpty`.

Reply:
If I added this check it wasn’t my best moment. :sweat_smile: 
Notice that if `resultsByExperiment` is empty, so will be the resulting stream and the lambda inside the collector won’t be executed.